File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
    const dfs = function (path, open, close) {
        if (path.length === 2 * n) {
            result.push(path.join(""));
            return;
        }

        if (open < n) {
            path.push('(');
            dfs(path, open + 1, close);
            path.pop();
        }

        if (close < open) {
            path.push(')');
            dfs(path, open, close + 1);
            path.pop();
        }
    }

    const result = [];
    dfs([], 0, 0);

    return result;
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def dfs(path, open, close):
            if len(path) == 2 * n:
                result.append("".join(path))
                return

            if open < n:
                path.append("(")
                dfs(path, open + 1, close)
                path.pop()

            if close < open:
                path.append(")")
                dfs(path, open, close + 1)
                path.pop()

        result = []
        dfs([], 0, 0)

        return result
